src/netlink.c

```diff
@@ -0,0 +1,385 @@
+static int meth_bind(lua_State *L);
+static int meth_sendto(lua_State *L);
+static int meth_send(lua_State *L);
+static int meth_receivefrom(lua_State *L);
+static int meth_receive(lua_State *L);
+static int meth_close(lua_State *L);
+static int meth_settimeout(lua_State *L);
+static int meth_gettimeout(lua_State *L);
+static int meth_getfd(lua_State *L);
+static int meth_setfd(lua_State *L);
+static int meth_getpeername(lua_State *L);
+static int meth_setpeername(lua_State *L);
+static int meth_getsockpid(lua_State *L);
+
+static const char *netlink_trybind(p_netlink nl, int grp);
+
+/* netlink object methods */
+static luaL_Reg netlink_methods[] = {
+    {"__gc",        meth_close},
+    {"__tostring",  auxiliar_tostring},
+    {"bind",        meth_bind},
+    {"close",       meth_close},
+    {"getfd",       meth_getfd},
+    {"send",        meth_send},
+    {"sendto",      meth_sendto},
+    {"receivefrom", meth_receivefrom},
+    {"receive",     meth_receive},
+    {"setfd",       meth_setfd},
+    {"settimeout",  meth_settimeout},
+    {"gettimeout",  meth_gettimeout},
+    {"setpeername", meth_setpeername},
+    {"getpeername", meth_getpeername},
+    {"getsockpid",  meth_getsockpid},
+    {NULL,          NULL}
+};
+
+/* functions in library namespace */
+static luaL_Reg func[] = {
+    {"netlink", global_create},
+    {NULL, NULL}
+};
+
+/*-------------------------------------------------------------------------*\
+* Initializes module
+\*-------------------------------------------------------------------------*/
+int netlink_open(lua_State *L) {
+    /* create classes */
+    auxiliar_newclass(L, "netlink{connected}", netlink_methods);
+    auxiliar_newclass(L, "netlink{unconnected}", netlink_methods);
+    /* create class groups */
+    auxiliar_add2group(L, "netlink{connected}", "netlink{any}");
+    auxiliar_add2group(L, "netlink{unconnected}", "netlink{any}");
+    luaL_setfuncs(L, func, 0);
+    return 0;
+}
+
+/*-------------------------------------------------------------------------*\
+* Send data through connected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_send(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int flags = luaL_optinteger(L, 3, 0);
+    struct nlmsghdr *nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    if (nlh == NULL) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error allocating memory in send()");
+        return 2;
+    }
+ 
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+
+    err = socket_send(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, tm);
```

Comment:
you are not marking timeout here like the original code does.. is this intentional?

Reply:
It is not and I will proceed to re-add that